hyper/http20/connection.py

```diff
@@ -525,7 +668,25 @@ def _recv_cb(self):
 
         This is generally called by a stream, not by the connection itself, and
         it's likely that streams will read a frame that doesn't belong to them.
+
+        :param stream_id: (optional) The stream ID of the stream reading data
+            from the connection.
+
         """
+        # Begin by reading what we can from the socket.
+        #
+        # Concurrency
+        #
+        # Ignore this read if some other thread has recently read data from
+        # from the requested stream
+        with self._read_lock:
+            log.debug('recv for stream %d with %s already present',
+                      stream_id,
+                      self.recent_recv_streams)
+            if stream_id in self.recent_recv_streams:
+                self.recent_recv_streams.discard(stream_id)
+                return
+
         # TODO: Re-evaluate this.
         self._single_read()
```

Comment:
Your analysis is spot on, thanks!   I opted for holding _read_lock across the whole method, aiming to keeping the number of locks involved to a minimum.